### PR TITLE
MULE-18655: Decorate source response streams with data flow statistic…

### PR DESCRIPTION
**Pull request by fsgonz**
…s decorator

### CODE REVIEW

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/source/DefaultSourceCallback.java [hunk 1]
```diff
@@ -249,14 +252,50 @@ public void handle(Result<T, A> result, SourceCallbackContext context) {
                                                                                      encodingParam,
                                                                                      mimeTypeInitParam);
 
+    T resultValue = result.getOutput();
+
+    Result<?, ?> decoratedResult = result;
+
+    if (resultValue instanceof InputStream) {
+      decoratedResult = decorateInputStream(result, context);
+    } else if (resultValue instanceof Iterator) {
```
Comment (elrodro83) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/source/DefaultSourceCallback.java:261]: this case must account PagingProviders rather than iterators? please check with the sdk guys

Comment (fsgonz) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/source/DefaultSourceCallback.java]: PagingProviders apply to operations.

**Comment by fsgonz** [comment]
--validate

**Comment by fsgonz** [comment]
--validate

**Comment by fsgonz** [comment]
--validate

**Comment by fsgonz** [comment]
--validate

File: core/src/main/java/org/mule/runtime/core/api/management/stats/CursorComponentDecoratorFactory.java [hunk 2]
```diff
@@ -67,6 +68,18 @@
...
+   * If statistics are enabled, decorates the provided {@link CursorStream}.
+   * <p>
+   * Ref: {@link PayloadStatistics#getInputByteCount()}.
+   *
+   * @param decorated the {@link CursorStream} to decorate.
+   * @param correlationId information to be used in the case a detailed report needs to be obtained, allowing to match the
+   *        measured volume to a specific execution.
+   * @return the decorated {@link CursorStream}.
+   */
+  CursorStream decorateInput(CursorStream decorated, String correlationId);
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/management/stats/CursorComponentDecoratorFactory.java:81]: why is this case different to the one with `InputSream`?

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/InputDecoratorVisitor.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    private boolean decorateIterators = true;
+    private boolean decorateInputStreams = true;
+    private boolean decorateCursorProviders = true;
+
+    private CursorComponentDecoratorFactory factory;
+    private String correlationId;
+
+    private Builder() {}
+
+    public Builder decorateCollections(boolean decorateCollections) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/InputDecoratorVisitor.java:115]: this is not used

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/InputDecoratorVisitor.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+      this.decorateCollections = decorateCollections;
+      return this;
+    }
+
+    public Builder decorateIterators(boolean decorateIterators) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/InputDecoratorVisitor.java:120]: this is not used

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/InputDecoratorVisitor.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+      this.decorateIterators = decorateIterators;
+      return this;
+    }
+
+    public Builder decorateInputStreams(boolean decorateInputStreams) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/InputDecoratorVisitor.java:125]: this is not used

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/NoOpCursorComponentDecoratorFactory.java [hunk 6]
```diff
@@ -67,4 +68,9 @@ public InputStream decorateOutput(InputStream decorated, String correlationId) {
   public Iterator<Result> decorateOutputResultIterator(Iterator<Result> decorated, String correlationId) {
     return decorated;
   }
+
+  @Override
+  public CursorStream decorateInput(CursorStream decorated, String correlationId) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/NoOpCursorComponentDecoratorFactory.java:73]: remove

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/OutputDecoratorVisitor.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+      this.decorateCollections = decorateCollections;
+      return this;
+    }
+
+    public Builder decorateIterators(boolean decorateIterators) {
+      this.decorateIterators = decorateIterators;
+      return this;
+    }
+
+    public Builder decorateInputStreams(boolean decorateInputStreams) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/OutputDecoratorVisitor.java:100]: this is not used

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/OutputDecoratorVisitor.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    private String correlationId;
+
+    private Builder() {}
+
+    public Builder decorateCollections(boolean decorateCollections) {
+      this.decorateCollections = decorateCollections;
+      return this;
+    }
+
+    public Builder decorateIterators(boolean decorateIterators) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/OutputDecoratorVisitor.java:95]: this is not used

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/OutputDecoratorVisitor.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    private boolean decorateIterators = true;
+    private boolean decorateInputStreams = true;
+    private boolean decorateCursorProviders = true;
+
+    private CursorComponentDecoratorFactory factory;
+    private String correlationId;
+
+    private Builder() {}
+
+    public Builder decorateCollections(boolean decorateCollections) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/OutputDecoratorVisitor.java:90]: this is not used

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/PayloadStatisticsCursorComponentDecoratorFactory.java [hunk 10]
```diff
@@ -145,4 +146,13 @@ private Object decorateOutput(Object decorated, String correlationId) {
     }
   }
 
+  @Override
+  public CursorStream decorateInput(CursorStream decorated, String correlationId) {
+    if (payloadStatistics.isEnabled()) {
+      return new PayloadStatisticsCursorStream(decorated, payloadStatistics::addInputByteCount);
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/PayloadStatisticsCursorComponentDecoratorFactory.java:152]: where is this class?

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/VisitableCollection.java [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+import java.util.Collection;
+import java.util.Iterator;
+
+/**
+ * A collection that can be visit to be decorated.
+ *
+ * @since 4.4, 4.3.1
+ */
+public class VisitableCollection implements Visitable<Collection>, Collection {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/VisitableCollection.java:17]: put all visitor stuff in its own subpackage

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/VisitableIterator.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+package org.mule.runtime.core.internal.management.stats;
+
+import java.util.Iterator;
+
+/**
+ * An iterator that can be visit to decorate.
+ *
+ * @since 4.4, 4.3.1
+ */
+public class VisitableIterator implements Iterator, Visitable<Iterator> {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/VisitableIterator.java:16]: this may need to implement `HasSize`

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/Visitor.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+   * @param visitable a visitable inputstream
+   * @return result of the inputstream which has been visited
+   */
+  InputStream visitInputStream(VisitableInputStream visitable);
+
+  /**
+   * @param visitable a visitable iterator
+   * @return result of the iterator which has been visited
+   */
+  Iterator visitIterator(VisitableIterator visible);
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/Visitor.java:32]: add generics to signature

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/Visitor.java [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+
+  /**
+   * @param visitable a visitable collection
+   * @return result of the collection which has been visited
+   */
+  Collection visitCollection(VisitableCollection visitable);
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/Visitor.java:38]: add generics to signature

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/Visitor.java [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+
+  /**
+   * @param visitable a visitable list
+   * @return result of the list which has been visited
+   */
+  List visitList(VisitableList visitableList);
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/Visitor.java:44]: add generics to signature

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/Visitor.java [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+
+  /**
+   * @param visitable a visitable set
+   * @return result of the set which has been visited
+   */
+  Set visitSet(VisitableSet visitableSet);
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/Visitor.java:50]: add generics to signature

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/objectbuilder/DefaultObjectBuilder.java [hunk 17]
```diff
@@ -127,27 +128,23 @@ public T build(ValueResolvingContext context) throws MuleException {
 
     for (Map.Entry<FieldSetter, ValueResolver<Object>> entry : resolvers.entrySet()) {
       final Object resolvedValue = resolveValue(entry.getValue(), context);
+      UnaryOperator decorateOperation = v -> {
+        return visitable(v)
```
Comment (elrodro83) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/objectbuilder/DefaultObjectBuilder.java:132]: avoid return for one-line lambdas

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/objectbuilder/ParameterGroupObjectBuilder.java [hunk 18]
```diff
@@ -113,15 +115,10 @@ private T doBuild(Predicate<String> hasParameter, Function<String, Object> param
...
-              if (v instanceof InputStream) {
-                return componentDecoratorFactory.decorateInput((InputStream) v, context.getEvent().getCorrelationId());
-              } else if (v instanceof Collection) {
-                return componentDecoratorFactory.decorateInput((Collection) v, context.getEvent().getCorrelationId());
-              } else if (v instanceof Iterator) {
-                return componentDecoratorFactory.decorateInput((Iterator) v, context.getEvent().getCorrelationId());
-              } else {
-                return v;
-              }
+              return visitable(v).map(visitable -> visitable
```
Comment (elrodro83) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/objectbuilder/ParameterGroupObjectBuilder.java:118]: avoid return for one-line lambdas

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/InputDecoratedCursorStreamProvider.java [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+
+  @Override
+  public CursorStream openCursor() {
+    return decoratorFactory.decorateInput(delegate.openCursor(), correlationId);
+  }
+
+  @Override
+  public void close() {
+    delegate.close();
+
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/InputDecoratedCursorStreamProvider.java:39]: remove empty line

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/InputDecoratedCursorStreamProvider.java [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+  }
+
+  @Override
+  public void releaseResources() {
+    delegate.releaseResources();
+
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/InputDecoratedCursorStreamProvider.java:45]: remove empty line

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/PayloadStatisticsCursorStream.java [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+ */
+package org.mule.runtime.core.internal.management.stats;
+
+import java.io.IOException;
+import java.util.function.LongConsumer;
+
+import org.mule.runtime.api.streaming.CursorProvider;
+import org.mule.runtime.api.streaming.bytes.CursorStream;
+
+final class PayloadStatisticsCursorStream extends CursorStream {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/PayloadStatisticsCursorStream.java:15]: why is this needed intead of using the inputstream decorator?

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/VisitableCursorStreamProvider.java [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+
+  @Override
+  public CursorStream openCursor() {
+    return null;
+  }
+
+  @Override
+  public void close() {
+    delegate.close();
+
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/VisitableCursorStreamProvider.java:33]: remove empty line

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/VisitableCursorStreamProvider.java [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  @Override
+  public void releaseResources() {
+    delegate.releaseResources();
+
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/VisitableCursorStreamProvider.java:39]: remove empty line

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/mule/runtime/core/api/management/stats/CursorComponentDecoratorFactory.java (hunk 2 above)
Comment (fsgonz) [reply core/src/main/java/org/mule/runtime/core/api/management/stats/CursorComponentDecoratorFactory.java:81]: When I want to decorate a CursorProvider on openCursor I need a CursorStream as a result. PayloadStatisticsInputStream extends InputStream, not CursorStream.
This is the wrapper for CursorStreamProvider which returns decorated cursors:
https://github.com/mulesoft/mule/pull/9338/files#diff-90fb1d0f6ef60883cd38dbf9b5bc8161R32

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/VisitableCursorStreamProvider.java [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+  private CursorStreamProvider delegate;
+
+  public VisitableCursorStreamProvider(CursorStreamProvider delegate) {
+    this.delegate = delegate;
+  }
+
+  @Override
+  public CursorStream openCursor() {
+    return null;
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/VisitableCursorStreamProvider.java:27]: this should delegate instead of returning null

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/NoOpCursorComponentDecoratorFactory.java (hunk 6 above)
Comment (fsgonz) [reply core/src/main/java/org/mule/runtime/core/internal/management/stats/NoOpCursorComponentDecoratorFactory.java:73]: Same commnent as above concerning CursorStream.

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/PayloadStatisticsCursorComponentDecoratorFactory.java (hunk 10 above)
Comment (fsgonz) [reply core/src/main/java/org/mule/runtime/core/internal/management/stats/PayloadStatisticsCursorComponentDecoratorFactory.java:152]: Already pushed (not originally)

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/PayloadStatisticsCursorStream.java (hunk 21 above)
Comment (fsgonz) [reply core/src/main/java/org/mule/runtime/core/internal/management/stats/PayloadStatisticsCursorStream.java:15]: same comment as before.

**Comment by fsgonz** [comment]
--validate

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/visitor/OutputDecoratorVisitor.java [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+    return visitableCursorStreamProvider;
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static class Builder {
+
+
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/visitor/OutputDecoratorVisitor.java:69]: remove extra line

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/visitor/VisitableCollection.java [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+  public void clear() {
+    delegate.clear();
+  }
+
+  @Override
+  public Collection<T> getDelegate() {
+    return delegate;
+  }
+
+
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/visitor/VisitableCollection.java:100]: remove extra lines

File: core/src/main/java/org/mule/runtime/core/internal/management/stats/visitor/VisitableList.java [hunk 27]
```diff
@@ -0,0 +1,153 @@
...
+  @Override
+  public List<T> subList(int fromIndex, int toIndex) {
+    return delegate.subList(fromIndex, toIndex);
+  }
+
+  @Override
+  public List<T> getDelegate() {
+    return delegate;
+  }
+
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/management/stats/visitor/VisitableList.java:151]: remove extra line

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ResolverUtils.java [hunk 28]
```diff
@@ -176,7 +192,8 @@ public static Object resolveCursor(TypedValue<?> typedValue, UnaryOperator value
       if (stackedTypesModelProperty.isPresent()) {
         resolver = stackedTypesModelProperty.get().getValueResolverFactory().getExpressionBasedValueResolver(expression,
                                                                                                              getType(type));
-        //TODO MULE-13518: Add support for stacked value resolvers for @Parameter inside pojos // The following "IFs" should be removed once implemented
+        // TODO MULE-13518: Add support for stacked value resolvers for @Parameter inside pojos // The following "IFs" should be
```
Comment (elrodro83) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ResolverUtils.java:195]: fix format, put the second `//` in a new line

**Review by elrodro83 — CHANGES_REQUESTED** [review]

**Comment by fsgonz** [comment]
--validate

**Comment by fsgonz** [comment]
--validate

**Review by elrodro83 — APPROVED** [review]

**Review by elrodro83 — APPROVED** [review]
